### PR TITLE
feat(react-components): change return value of use3dScenes

### PR DESCRIPTION
**Pull request by anders-hopland**
#### Type of change
Refactor 

## Description :pencil:
Changing return value of `use3dScenes` as we also need the name of a given Scene. By returning a custom type we can add more properties the custom type without creating a breaking change in the future

## Checklist :ballot_box_with_check:

- [x] I am proud of this feature.
- [x] I have performed a self-review of my own code.
- [x] I have added PR type (Feat, Bug, Chore, Test, Docs, Style, Refactor) to the PR title.
- [ ] I have manually tested this for different scenarios (different models, formats, devices, browsers).
- [ ] I have commented my code in hard-to-understand areas.
- [ ] I have made corresponding changes to the public documentation.
- [ ] I have added unit and visuals tests to prove that my feature works to the best of my ability.
- [ ] I have refactored the code for readability to the best of my ability.
- [ ] I have checked that my changes do not introduce regressions in the public documentation.
- [ ] I have outlined any known defects / lacking capabilities in the contents of this PR.
- [ ] I have listed any remaining work that should follow this PR in the description or jira/miro/etc.
- [ ] I have added TSDoc to any public facing changes.
- [ ] I have added "developer documentation" in any package README.md that is related / applicable to this PR.
- [ ] I have noted down and am currently tracking any technical debt introduced in this PR.
- [ ] I have thoroughly thought about the architecture of this implementation.
- [ ] I have asked peers to test this feature.

### CODE REVIEW

File: react-components/src/hooks/use3dScenes.tsx [hunk 1]
```diff
@@ -58,25 +61,36 @@ export const use3dScenes = (
...
-            spaceMap[externalId] = [];
+            spaceMap[externalId] = { name: '', modelOptions: [] };
           }
           const externalIdMap = spaceMap[externalId];
-          externalIdMap.push(newModel);
+          externalIdMap.modelOptions.push(newModel);
 
           return acc;
         },
         // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:71]: What is this needed for?

File: react-components/src/hooks/use3dScenes.tsx [hunk 2]
```diff
@@ -58,25 +61,36 @@ export const use3dScenes = (
...
+      scenesQueryResult.items.scenes.forEach((scene) => {
+        const { space, externalId } = scene;
+
+        if (scenesMap[space]?.[externalId] !== undefined) {
+          const properties = Object.values(
+            Object.values(scene.properties)[0] as Record<string, unknown>
+          )[0] as SceneConfigurationProperties;
+          scenesMap[space][externalId].name = properties.name;
+        }
+      });
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:84]: I would like this to be put in its own function and be done before the `reduce` above.. In other words - this block creates the map from space-externalId to scene object with name, and the reduce-block fills it with models

**Review by haakonflatval-cognite — CHANGES_REQUESTED** [review]
Some small comments - otherwise looks good :+1:

File: react-components/src/hooks/use3dScenes.tsx (hunk 1 above)
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx:71]: We use the `modelOptions` in use3dScenes for finding mapped objects for a given scene in other applications

File: react-components/src/hooks/use3dScenes.tsx (hunk 2 above)
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx]: Fixed in [d11b2b7](https://github.com/cognitedata/reveal/pull/4131/commits/d11b2b779ae03f129537389b15adf9d446582de0) along with some refactoring

File: react-components/src/hooks/use3dScenes.tsx [hunk 3]
```diff
@@ -33,56 +40,11 @@ export const use3dScenes = (
...
-            Object.values(scene.properties)[0] as Record<string, unknown>
-          )[0] as SceneConfigurationProperties;
-          scenesMap[space][externalId].name = properties.name;
-        }
-      });
+      const scenesMap = createMapOfScenes(scenesQueryResult.items.scenes);
+      populateSceneMapWithModels(scenesQueryResult.items.sceneModels, scenesMap);
+      populateSceneMapWith360Images(scenesQueryResult.items.scene360Collections, scenesMap);
 
+      console.log('scenesMap', scenesMap);
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:47]: :eyes:

File: react-components/src/hooks/use3dScenes.tsx [hunk 4]
```diff
@@ -96,11 +58,99 @@ export const use3dScenes = (
...
+        acc[space][externalId] = {
+          name: properties.name,
+          cadModelOptions: [],
+          image360CollectionOptions: []
+        };
+      }
+
+      return acc;
+    },
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:83]: I think this comment can be removed? :thinking:

File: react-components/src/hooks/use3dScenes.tsx [hunk 5]
```diff
@@ -96,11 +58,99 @@ export const use3dScenes = (
   );
 };
 
-function createModelFromEdge(
-  newModelId: number,
-  newModelRevisionId: number,
-  properties: Cdf3dRevisionProperties
-): AddReveal3DModelOptions {
+function createMapOfScenes(
+  scenes: Array<NodeItem<Record<string, unknown>>> | Array<EdgeItem<Record<string, unknown>>>
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:62]: Can this be just `Array<EdgeItem<Record<string, Record<string, Cdf3dRevisionProperties>>>>`? I want to get rid of as many `as`-es as possible

File: react-components/src/hooks/use3dScenes.tsx [hunk 6]
```diff
@@ -96,11 +58,99 @@ export const use3dScenes = (
...
+    },
+    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+    {} as Record<Space, Record<ExternalId, SceneData>>
+  );
+}
+
+function populateSceneMapWithModels(
+  scene360Images:
+    | Array<NodeItem<Record<string, unknown>>>
+    | Array<EdgeItem<Record<string, unknown>>>,
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:91]: Same here, can we be more specific here and get rid of the `as`-es? :sweat_smile:

File: react-components/src/hooks/use3dScenes.tsx [hunk 7]
```diff
@@ -96,11 +58,99 @@ export const use3dScenes = (
...
+    };
+
+    scenesMap[space]?.[externalId].cadModelOptions.push(newModel);
+  });
+}
+
+function populateSceneMapWith360Images(
+  scene360Images:
+    | Array<NodeItem<Record<string, unknown>>>
+    | Array<EdgeItem<Record<string, unknown>>>,
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/use3dScenes.tsx:128]: More specific type and less `as`-es?

**Review by haakonflatval-cognite — CHANGES_REQUESTED** [review]
Looks better - still some comments

File: react-components/src/hooks/use3dScenes.tsx [hunk 8]
```diff
@@ -184,6 +230,23 @@ function createGetScenesQuery(limit: number = 100): Query {
...
+              property: ['edge', 'type'],
+              value: {
+                space: 'scene',
+                externalId: 'SceneConfiguration.images360Collections'
+              }
+            }
+          }
+        },
+        limit
       }
```
[the remark this answers is not in the document]
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx:250]: Correct!

File: react-components/src/hooks/use3dScenes.tsx (hunk 3 above)
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx]: Good catch 😅 , fixed in [808e42b](https://github.com/cognitedata/reveal/pull/4131/commits/808e42b35074ad506aac0d3f47f0e32ad69050fa)

File: react-components/src/hooks/use3dScenes.tsx (hunk 5 above)
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx]: No, unfortunately not. Changed it to `NodeItem[] | EdgeItem[]` in [808e42b](https://github.com/cognitedata/reveal/pull/4131/commits/808e42b35074ad506aac0d3f47f0e32ad69050fa)

File: react-components/src/hooks/use3dScenes.tsx (hunk 6 above)
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx]: No, unfortunately not. Changed it to NodeItem[] | EdgeItem[] in [808e42b](https://github.com/cognitedata/reveal/pull/4131/commits/808e42b35074ad506aac0d3f47f0e32ad69050fa)

File: react-components/src/hooks/use3dScenes.tsx (hunk 7 above)
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx]: No, unfortunately not. Changed it to NodeItem[] | EdgeItem[] in [808e42b](https://github.com/cognitedata/reveal/pull/4131/commits/808e42b35074ad506aac0d3f47f0e32ad69050fa)

File: react-components/src/hooks/use3dScenes.tsx [hunk 9]
```diff
@@ -96,11 +58,99 @@ export const use3dScenes = (
...
+  scenesMap: Record<Space, Record<ExternalId, SceneData>>
+): void {
+  scene360Images.forEach((imageCollection) => {
+    const edge = imageCollection as EdgeItem;
+
+    const { space, externalId } = edge.startNode;
+
+    const properties = Object.values(
+      Object.values(edge.properties)[0] as Record<string, unknown>
+    )[0] as Cdf3dImage360CollectionProperties;
```
[the remark this answers is not in the document]
Comment (anders-hopland) [reply react-components/src/hooks/use3dScenes.tsx:132]: With only 3 occurences I don't think that it's worth it to abstract it away

**Review by haakonflatval-cognite — APPROVED** [review]
LGTM :+1:

**Pull request merged**